### Rudraksh Sindwani                          SIT720


# Part B: Problem Solving Practical Task

## Question 1: Linear SVM

### Discussion

The Bank Marketing dataset was loaded to predict whether a customer subscribed to a term deposit. The target variable `y` has two classes: yes and no.

The dataset contains numerical and categorical variables. The categorical variables were changed into numerical values using One-Hot Encoding. The data was divided into training and testing sets using stratified sampling because the target classes are not balanced. Standard Scaling was then applied to the features.

A Support Vector Machine with a linear kernel and the default value of `C = 1.0` was used. Accuracy precision recall and F1-score were used to measure the model performance.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

bank = pd.read_csv("bank-additional-full.csv", sep=";")
print("\nTarget variable:")
print(bank["y"].value_counts())


Target variable:
y
no     36548
yes     4640
Name: count, dtype: int64


In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Separate features and target
X = bank.drop("y", axis=1)
y = bank["y"].map({"no": 0, "yes": 1})

# One-Hot Encoding
Xencoded = pd.get_dummies(X, drop_first=True)

print("Features after encoding:", Xencoded.shape[1])

# Split data into training and testing sets
Xtrain, Xtest, ytrain, ytest = train_test_split(
    Xencoded,
    y,
    test_size=0.2,
    random_state=0,
    stratify=y
)

print("Training samples:", Xtrain.shape[0])
print("Testing samples:", Xtest.shape[0])


# Standard Scaling
sc = StandardScaler()
sc.fit(Xtrain)

X_train_std = sc.transform(Xtrain)
X_test_std = sc.transform(Xtest)

print("Scaled training shape:", X_train_std.shape)
print("Scaled testing shape:", X_test_std.shape)

Features after encoding: 53
Training samples: 32950
Testing samples: 8238
Scaled training shape: (32950, 53)
Scaled testing shape: (8238, 53)


In [9]:
from sklearn.svm import SVC
from sklearn.metrics import (accuracy_score,precision_score,recall_score,f1_score,confusion_matrix)

# Linear SVM with default C = 1.0
svm_linear = SVC(kernel="linear",C=1.0,random_state=0)

# Train model
svm_linear.fit(X_train_std, ytrain)

# Predictions
ypred = svm_linear.predict(X_test_std)

# Performance metrics
accuracy = accuracy_score(ytest, ypred)
precision = precision_score(ytest, ypred)
recall = recall_score(ytest, ypred)
f1 = f1_score(ytest, ypred)

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

print("\nConfusion Matrix:")
print(confusion_matrix(ytest, ypred))

Accuracy: 0.9031318281136198
Precision: 0.6504629629629629
Recall: 0.30280172413793105
F1-score: 0.41323529411764703

Confusion Matrix:
[[7159  151]
 [ 647  281]]


## Result

The Linear SVM model with the default C = 1.0 achieved an accuracy of 90.31%. The precision was 65.05% and the recall was 30.28%. The F1-score was 41.32%.

The confusion matrix shows that the model correctly classified 7159 customers who did not subscribe and 281 customers who subscribed. However 647 customers who subscribed were incorrectly classified as not subscribing.

The model has high accuracy but low recall. This happens because the dataset has many more customers who did not subscribe than customers who subscribed. The model therefore needs further improvement for detecting the smaller class.